refactor(rxNotify): components/rxNotify to elements/rxNotify

khaziwallis opened the pull request:
JIRA: https://jira.rax.io/browse/EO-1084

### LGTMs
- [x] Dev LGTM
- [x] Dev+Vidyo LGTM
- [x] Design LGTM
- [x] QE LGTM


Before
---------
Components > rxNotify
---------
![screen shot 2016-07-21 at 8 05 08 pm](https://cloud.githubusercontent.com/assets/3723469/17026671/804d1404-4f7e-11e6-87d2-61aeb78f2552.png)

After
---------
Elements > Notifications
---------
Simple Notification
---------
![screen shot 2016-07-22 at 5 11 07 pm](https://cloud.githubusercontent.com/assets/3723469/17056189/57828de0-502f-11e6-8287-bb6c8cd688b6.png)


Stack Notification
----------
![screen shot 2016-07-22 at 5 03 48 pm](https://cloud.githubusercontent.com/assets/3723469/17056043/76e446c0-502e-11e6-8393-9a9c2757e9fc.png)

promise notification
---------
![screen shot 2016-08-01 at 8 27 58 pm](https://cloud.githubusercontent.com/assets/3723469/17298162/742dcc58-5826-11e6-89eb-32d7c3ab8d16.png)


Custom stack notification
---------
![screen shot 2016-08-01 at 8 27 51 pm](https://cloud.githubusercontent.com/assets/3723469/17298179/89d88b56-5826-11e6-888a-3d2905cb6f87.png)


route notification
---------
![screen shot 2016-07-22 at 5 03 57 pm](https://cloud.githubusercontent.com/assets/3723469/17056070/a2716ee4-502e-11e6-962a-140152b5adc6.png)

Full Page
---------
Elements > Notifications
---------
![screen shot 2016-07-22 at 5 10 50 pm](https://cloud.githubusercontent.com/assets/3723469/17056198/67159e0a-502f-11e6-805f-d231a87c4893.png)





When view All Elements/Molecules
---------
 And filter by "notifications"
---------
![screen shot 2016-07-21 at 8 07 53 pm](https://cloud.githubusercontent.com/assets/3723469/17026844/2147582e-4f7f-11e6-9155-f4a2e81046ef.png)
![screen shot 2016-07-21 at 8 12 29 pm](https://cloud.githubusercontent.com/assets/3723469/17026988/ab56b0be-4f7f-11e6-8134-0a193ffb7433.png)



 And filter by "notification"
---------
![screen shot 2016-07-21 at 8 07 47 pm](https://cloud.githubusercontent.com/assets/3723469/17026871/3cf7e55c-4f7f-11e6-8c8f-d013dc1df424.png)
![screen shot 2016-07-21 at 8 14 32 pm](https://cloud.githubusercontent.com/assets/3723469/17027015/d0833eb6-4f7f-11e6-804f-75dd0fd21555.png)


 And filter by "error"
---------

![screen shot 2016-07-21 at 8 08 34 pm](https://cloud.githubusercontent.com/assets/3723469/17026880/466a0002-4f7f-11e6-80fb-1121ba24f7d6.png)

![screen shot 2016-07-21 at 8 12 29 pm](https://cloud.githubusercontent.com/assets/3723469/17026954/8cadcf76-4f7f-11e6-810e-34a3eabd4506.png)


 And filter by "help"
---------
![screen shot 2016-07-21 at 8 07 38 pm](https://cloud.githubusercontent.com/assets/3723469/17026900/586ca7fa-4f7f-11e6-811a-f42d19615cc2.png)

![screen shot 2016-07-21 at 8 14 38 pm](https://cloud.githubusercontent.com/assets/3723469/17027020/d8681502-4f7f-11e6-97c9-e5f719a688cb.png)

Droogans commented:
I believe this should be under Elements -> Notifications.

We are moving away from using the component/directive name in the documentation headers.

khaziwallis commented:
@Droogans  thanks for solving the file case issue and helping on PR conventions. please take a look for review.

100stacks commented:
@khaziwallis, your screenshots are using the deprecated demo pattern.  They should be using `rxExample`.

Also, what is important is the view of the entire demo page, not just the left nav panel.

100stacks commented on src/elements/Notifications/notifications.meta.json:
```diff
@@ -2,7 +2,6 @@
     "displayName": "Notifications",
     "stability": "stable",
     "description": "Manages page messages for an application.",
-    "isLegacy": true,
```
Change from:
```js
 "isLegacy": true,
```
to:
```js
"api": "directive:rxNotification",
```

This change is what tells the build process to use the `rxExample` demo styles.

khaziwallis commented:
@Droogans  facing some error in notifications.spec.js file `Attempting to use an unsafe value in a safe context.`. can u suggest me ways to solve this.

100stacks commented on src/components/rxApp/scripts/rxApp.spec.js:
```diff
@@ -1,4 +1,9 @@
+
+    beforeEach(function () {
+        module('templates/rxFeedback.html');
+    });
```
@khaziwallis: What is the reason for this update?

khaziwallis replied:
@100stacks was getting issue at unit test. please find attached error as screen shot.
![screen shot 2016-07-21 at 11 41 50 pm](https://cloud.githubusercontent.com/assets/3723469/17033690/f2d03d62-4f9c-11e6-995a-12018b0cfd54.png)

100stacks commented on src/elements/Notifications/scripts/rxNotification.spec.js:
```diff
@@ -0,0 +1,46 @@
+    var messageText1 = 'My Message 1';
+    var defaultStack = 'page';
+    beforeEach(function () {
+        module('encore.ui.elements');
```
👍

100stacks commented on src/elements/Notifications/scripts/rxNotification.spec.js:
```diff
@@ -0,0 +1,46 @@
+            compile = $compile;
+            notifySvc = rxNotify;
+        });
+        var rxNotificationTemplate = '<rx-notification type="warning">' + messageText1 + '</rx-notification>';
```
@khaziwallis: You removed the `describe` block:

```js
 describe('Directive: rxNotification', function () {
        beforeEach(function () {
            var rxNotificationTemplate = '<rx-notification type="warning">' + messageText1 + '</rx-notification>';
            el2 = helpers.createDirective(rxNotificationTemplate, compile, scope);
        });
```

Please revert.

100stacks commented on src/elements/Notifications/scripts/rxNotification.spec.js:
```diff
@@ -0,0 +1,46 @@
+describe('Directive: rxNotification', function () {
```
Change from
```js
describe('Directive: rxNotification', function () {
```
to:
```js
 
describe('Element: Notifications', function () {
```

khaziwallis commented:
@100stacks have addressed the review comments. need suggestion on unit test issues. please take a look on screen shot below.
![screen shot 2016-07-21 at 11 53 28 pm](https://cloud.githubusercontent.com/assets/3723469/17034028/6eb09c5a-4f9e-11e6-8c85-361a94e803a8.png)

glynnis commented:
@khaziwallis We should have specified this in the JIRA story, and we will look through other stories we've assigned to see if there's more detail like this we can add so that it's clear.

My only design request is that we split the demos up so they're more granular and so it's easier for users to understand and find code quickly for a specific demo they're looking at, a lot like how we've done on the Tables page (http://rackerlabs.github.io/encore-ui/#/elements/Tables) where there are lots of different demos.

Could you break the demos and their code up as follows into 5 different demos, each with their own section of code?
![screen_shot_2016-07-21_at_2_35_03_pm](https://cloud.githubusercontent.com/assets/1529366/17036217/f8e463b4-4f50-11e6-85b3-4d305c5126f8.png)
![screen_shot_2016-07-21_at_2_35_11_pm](https://cloud.githubusercontent.com/assets/1529366/17036218/f8e87a4e-4f50-11e6-97cd-5551376a02dc.png)

Thanks! Let me know if I can clarify anything here.

khaziwallis commented:
@glynnis  have updated the demo examples. as per suggestion. please take a look on stack example, i felt appropriate to keep 1 and 2 in single example.

glynnis commented:
Awesome! This LGTM from a design standpoint.

khaziwallis commented:
@Droogans  @100stacks  jenking getting stop on midways run. its working fine on my local.

tomreece commented:
@khaziwallis asked me to help remove the first 2 commits (by Satyam, from other branches). I did that on this branch: https://github.com/rackerlabs/encore-ui/compare/EO-1084-refactor-notification-cherry-pick

@khaziwallis you should look over the EO-1084-refactor-notification-cherry-pick branch and make sure everything is as you expect.

Droogans commented:
@tomreece Sure thing. I'm fixing an alert dialog not expected error in protractor for this PR.

Once it's fixed I'll resquash it all back in and push it up on top of this open PR. Thanks!

khaziwallis commented:
@100stacks please take a look for review.

CITguy commented on src/elements/AccountInfo/scripts/rxAccountInfo.spec.js:
```diff
@@ -75,7 +75,7 @@ describe('elements:rxAccountInfo', function () {
 
             // load module
-            module('encore.ui.rxNotify');
+            module('encore.ui.elements');
```
This line is no longer necessary, given that there is a similar line below.

CITguy commented on src/elements/Notifications/docs/Notifications.docs.html:
```diff
@@ -0,0 +1,15 @@
+<p>
+   The Notifications component provides status message notifications on a page.
```
Per the [HTML Styleguide](https://github.com/rackerlabs/encore-ui/blob/master/guides/html-styleguide.md#use-two-space-indentations), please update to use 2 spaces per indent.

CITguy commented on src/elements/Notifications/docs/examples/notifications.promise.html:
```diff
@@ -0,0 +1,8 @@
+<div>
+    <div>
```
Per the [HTML Styleguide](https://github.com/rackerlabs/encore-ui/blob/master/guides/html-styleguide.md#use-two-space-indentations), please update file to use 2 spaces per indent.

CITguy commented on src/elements/Notifications/docs/examples/notifications.route.html:
```diff
@@ -0,0 +1,8 @@
+<div ng-controller="notificationsRouteCtrl">
+    <div>
```
Per the [HTML Styleguide](https://github.com/rackerlabs/encore-ui/blob/master/guides/html-styleguide.md#use-two-space-indentations), please update file to use 2 spaces per indent.

CITguy replied:
* Extra `<div>` elements are unnecessary and distracting for user. 
* `<h3>` is unnecessary.

Please update content to the following:

```html
<div ng-controller="notificationsRouteCtrl">
  <button class="button" ng-click="routeChange()">Simulate Route Change</button>
</div>
```

CITguy commented on src/elements/Notifications/docs/examples/notifications.simple.html:
```diff
@@ -0,0 +1,22 @@
+<div>
+    <rx-notification type="error">Hello, world! <a href="#">This is a link.</a></rx-notification>
```
Per the [HTML Styleguide](https://github.com/rackerlabs/encore-ui/blob/master/guides/html-styleguide.md#use-two-space-indentations), please update file to use 2 spaces per indent.

CITguy commented on src/elements/Notifications/docs/Notifications.docs.html:
```diff
@@ -0,0 +1,15 @@
+<rx-example name="notifications.stack"></rx-example>
+
+<h2><rx-permalink>Promise Notifications</rx-permalink></h2>
+<rx-example name="notifications.promise"></rx-example>
```
Please move static HTML content out of `notifications.promise.html` back into this page. The example html provides no useful source code for rxExample purposes.

CITguy commented on src/elements/Notifications/docs/examples/notifications.promise.html:
```diff
@@ -0,0 +1,8 @@
+          <p>Please see <a href="#/utilities/rxPromiseNotifications">rxPromiseNotifications</a> demo.</p>
+        </div>
+    </div>
+</div>
```
Please move content back to `src/elements/Notifications/docs/Notifications.docs.html`

CITguy commented on src/elements/Notifications/docs/examples/notifications.simple.html:
```diff
@@ -0,0 +1,22 @@
+    </div>
+
+    <rx-notification stack="demo-stack" type="info">Hello, world in demo-stack stack!</rx-notification>
+    <rx-notification stack="custom-stack" type="warning">Hello, world in custom-stack stack!</rx-notification>
```
This example is too large and ambiguous. Lines 7-21 should be moved to a `notifications.custom-stack.html` to provide a more focused example.

CITguy commented on src/elements/Notifications/docs/examples/notifications.simple.html:
```diff
@@ -0,0 +1,22 @@
+
+    <h3>Using rxNotification with custom stack</h3>
+    <div layout="row">
+    <div flex>
```
Nested indentation needs corrected. Please indent this `<div>` and its siblings one more level.

CITguy commented:
@glynnis: I'd recommend splitting section 5 in your hand-drawn screenshot above. Simple notifications and custom stack notifications are slightly different with regards to their implementation.

khaziwallis commented:
@CITguy  have made changes as per review comments. please take a look.

CITguy commented on src/elements/Notifications/docs/Notifications.docs.html:
```diff
@@ -0,0 +1,21 @@
+
+<h2><rx-permalink>Promise Notifications</rx-permalink></h2>
+<div>
+  <h3>rxPromiseNotifications</h3>
```
This heading may be unnecessary. Please remove this `<h3>`

CITguy commented on src/elements/Notifications/docs/Notifications.docs.html:
```diff
@@ -0,0 +1,21 @@
+<h2><rx-permalink>Stack Notifications</rx-permalink></h2>
+<rx-example name="notifications.stack"></rx-example>
+
+<h2><rx-permalink>Custom Stack Notifications</rx-permalink></h2>
```
Please move "Custom Stack Notifications" between "Stack Notifications" and "Simple Notifications"

CITguy commented on src/elements/Notifications/docs/examples/notifications.simple.html:
```diff
@@ -0,0 +1,6 @@
+  <rx-notification type="warning">Hello, world! <a href="#">This is a link.</a></rx-notification>
+  <rx-notification type="info">Hello, world! <a href="#">This is a link.</a></rx-notification>
+  <rx-notification type="success">Hello, world! <a href="#">This is a link.</a></rx-notification>
+</div>
```
`<div>` is unnecessary. Please remove.

CITguy commented on src/elements/Notifications/docs/examples/notifications.custom-stack.html:
```diff
@@ -0,0 +1,15 @@
+  </div>
+  <rx-notification stack="demo-stack" type="info">Hello, world in demo-stack stack!</rx-notification>
+  <rx-notification stack="custom-stack" type="warning">Hello, world in custom-stack stack!</rx-notification>
+</div>
```
`<div>` is unnecessary. Please remove.

CITguy commented on src/elements/Notifications/docs/Notifications.docs.html:
```diff
@@ -0,0 +1,21 @@
+<div>
+  <h3>rxPromiseNotifications</h3>
+  <p>Please see <a href="#/utilities/rxPromiseNotifications">rxPromiseNotifications</a> demo.</p>
+</div>
```
`<div>` is unnecessary. Please remove.

khaziwallis commented:
@CITguy  have addressed review comments. please take a look.

100stacks commented:
@khaziwallis @satyamkotakonda, please remove the label after you have addressed the comments.

<img width="210" alt="screen shot 2016-08-02 at 3 55 48 pm" src="https://cloud.githubusercontent.com/assets/10120600/17344927/8645ae08-58c9-11e6-8c95-30f832e6991d.png">

This will help reviewers quickly see that the PR is ready for review.  👍

100stacks commented on src/components/rxNotify/rxNotify.module.js:
```diff
@@ -1,82 +0,0 @@
- */
-angular.module('encore.ui.rxNotify', [
-    'ngSanitize',
-    'ngAnimate',
```
@khaziwallis @satyamkotakonda, this module has an `ngAnimate` dependency.

CITguy replied:
@100stacks: the `encore.ui.elements` module already has `ngAnimate` as a dependency

100stacks replied:
@CITguy, I referenced it here to show @khaziwallis @satyamkotakonda @manojdas2212 where component dependency injections are located...in case they come across it again on Element migrations.  👍

100stacks commented on src/elements/elements.module.js:
```diff
@@ -16,5 +16,6 @@
 angular.module('encore.ui.elements', [
-    'encore.ui.utilities'
+    'encore.ui.utilities',
+    'ngSanitize'
```
@khaziwallis @satyamkotakonda, please add `ngAnimate` dependency to `src/elements/elements.module.js`.

khaziwallis commented:
@100stacks  @CITguy  have addressed review comments. please take a look.

CITguy commented on src/elements/Notifications/scripts/rxNotification.js:
```diff
@@ -1,4 +1,4 @@
+angular.module('encore.ui.elements')
 /**
  * @ngdoc directive
  * @name rxNotify.directive:rxNotification
```
Name should be `@name elements.directive:rxNotification`

CITguy commented on src/elements/Notifications/scripts/rxNotifications.js:
```diff
@@ -1,4 +1,4 @@
+angular.module('encore.ui.elements')
 /**
  * @ngdoc directive
  * @name rxNotify.directive:rxNotifications
```
Name should be `@name elements.directive:rxNotifications`

CITguy commented on src/components/rxNotify/rxNotify.module.js:
```diff
@@ -1,82 +0,0 @@
- * <pre>
- * <rx-notifications stack="loginForm"></rx-notifications>
- * </pre>
- */
```
Lines 7-76 need migrated to documentation in `src/utilities/rxNotify/scripts/rxNotify.js`

CITguy commented:
There's a warning in console when generating ngDocs. Please correct.

![screen shot 2016-08-03 at 1 16 09 pm](https://cloud.githubusercontent.com/assets/545605/17376757/96814bf2-597c-11e6-9555-4fcaf7eb3855.png)

CITguy commented:
LGTM
